### PR TITLE
Using Valiant for Validations in Agendum: A Dive into DBIx::Class Integration

### PR DESCRIPTION
**Pull request by jjn1056**
Article for issue https://github.com/perladvent/Perl-Advent/issues/484

### CODE REVIEW

**Comment by oalders** [comment]
Thank you for your contribution! I'm finding that blog articles which include images have much more compelling social media previews when the article is shared. Adding an image is by no means required and you should feel free to ignore this message entirely. However, if you'd like to add an image, it's as simple as:

* Adding the image to the `2024/share/static` folder like `2024/share/static/some-graphic.png`
* Including the image as HTML in your article

The HTML would look like:

```
=for :html
<img src="some-graphic.png">

```

The trailing blank line is important as that signals that the `HTML` section is finished.

Images can most certainly be AI generated. You should also feel free to use https://search.creativecommons.org/ in order to find CC images which can be freely shared. If you choose to go this route, please provide proper attribution when including the image, if attribution is required.

Most importantly, if you don't feel like doing this, then feel free to submit your article without an image. It's *your* article and *you* need to be happy with it.

**Comment by jjn1056** [comment]
@oalders you inspired me to make a logo for Valiant, please let me know if I followed your instructions correctly :)

**Comment by oalders** [comment]
This is great! Do you have any peers you'd like to have review this? If not, that's fine, but I'm finding I don't have the resources to do a really deep dive on articles right now. I'm happy to fix up formatting issues etc.

**Comment by jjn1056** [comment]
@oalders I did ask three people to look at it but didn't get a lot of feedback.   I worry a bit that I'm so close to this stuff that I'm not writing something people unfamiliar with it can understand but I get everyone is busy right now.  If you look it over and feel lost let me know.   I ran it thru two grammar checkers so hopefully you won't have a ton of work on that score.

File: 2024/incoming/using-valiant-for-validations-in-agendum-a-deep-dive-into-dbix-class-integration.pod [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+    my %errors = $user->errors->to_hash(full_messages=>1);
+
+This will return a hash:
+  
+    ( email => ['Email is not an email address'] )
+
+There's a lot of different ways to define validations, to customize error messages, and to retrieve the
+errors.  You can also define custom validators, and you can define validations that are conditional on
+other attributes.  You can also define validations that are conditional on the object state.  All this
+is covered in the full L<Valiant> documentation. But this gives you the basic idea.
```
Comment (openstrike) [inline 2024/incoming/using-valiant-for-validations-in-agendum-a-deep-dive-into-dbix-class-integration.pod:74]:
```suggestion
is covered in the full L<Valiant> documentation but this gives you the basic idea.
```

**Review by openstrike — CHANGES_REQUESTED** [review]
Thanks for this article. I found it very interesting and the level (for me) was just right. I've made quite a few suggestions here but they are all very minor - typos and so forth.

In case you were unaware, the badges on the metacpan page for Valiant show failures in the Linux and Windows test/coverage. I only mention because this will be the first time a lot of readers will look at it and you'll probably want to give them the best possible first impression.

File: 2024/incoming/using-valiant-for-validations-in-agendum-a-deep-dive-into-dbix-class-integration.pod [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+=head2 Conclusion
+
+I hope this article has given you a basic overview of how to use L<Valiant> with L<DBIx::Class>.  I skipped
+on a few of the more advanced features, such as custom validators, delegating validation to other objects,
+internationalization, filtering, using the form builder and more.  If you want to get a feel 
+for some of those things you can play with the Agendum application and read more about them in the full 
+L<Valiant> documentation.
+
+I'm always looking for feedback on L<Valiant> and the L<DBIx::Class> integration.  If you have any thoughts
+or suggestions please feel free to reach out to me via email or on the L<Valiant> github repository.  I'm
```
[the remark this answers is not in the document]
Comment (jjn1056) [reply 2024/incoming/using-valiant-for-validations-in-agendum-a-deep-dive-into-dbix-class-integration.pod:500]: good call, I'll update

**Comment by jjn1056** [comment]
> Thanks for this article. I found it very interesting and the level (for me) was just right. I've made quite a few suggestions here but they are all very minor - typos and so forth.
> 
> In case you were unaware, the badges on the metacpan page for Valiant show failures in the Linux and Windows test/coverage. I only mention because this will be the first time a lot of readers will look at it and you'll probably want to give them the best possible first impression.

@openstrike thanks for looking thru this.   Even after reading it 10 times there's still so many grammar errors, sorry about that:(.  

I did a new release of Valiant to remove the windows tests results (I can't seem to get at action to work right no matter what) and tweaked the actions to work properly again for Mac and Linux.  So that should look like not abandoned code on Metacpan when people see it.

Thanks for the feedback.   It would be nice to get some people excited about this, I use it on a lot of projects but it would be better to not be the only one.

**Review by openstrike — APPROVED** [review]

**Comment by oalders** [comment]
> @oalders I'm happy for this to be merged but maybe you'd like to have a quick read-through for good measure?

I trust you. 😄 Thank you so much for your help with this @openstrike!

Thanks @jjn1056 for sharing your hard work!

**Pull request merged**